src/daft-minhash/src/lib.rs

```diff
@@ -1,6 +1,703 @@
+/// +-----+-----+-----+-----+-----+-----+-----+-----+
+///
+///        Rotate Hash Values Left
+/// [H8]   [H1]  [H2]  [H3]  [H4]  [H5]  [H6]  [H7]
+///   |      |     |     |     |     |    |     |
+///   v      v     v     v     v     v    v     v
+/// +-----+-----+-----+-----+-----+-----+-----+-----+
+/// | P1  | P2  | P3  | P4  | P5  | P6  | P7  | P8  |  Permutation Sets
+/// +-----+-----+-----+-----+-----+-----+-----+-----+
+///   |     |     |     |     |     |     |     |
+///   v     v     v     v     v     v     v     v
+/// +-----+-----+-----+-----+-----+-----+-----+-----+
+/// | M1  | M2  | M3  | M4  | M5  | M6  | M7  | M8  |  Min Hash Values
+/// +-----+-----+-----+-----+-----+-----+-----+-----+
+///   ^     ^     ^     ^     ^     ^     ^     ^
+///   |     |     |     |     |     |     |     |
+///   |     |     |     |     |     |     |     |
+///   +-----+-----+-----+-----+-----+-----+-----+
+///    (Update with minimum of new and existing values)
+///
+///        Rotate Hash Values Left
+///   [H7]  [H8]  [H1]  [H2]  [H3]  [H4] [H5]   [H6]
+///     |    |     |     |     |     |    |      |
+///     v    v     v     v     v     v    v      v
+/// +-----+-----+-----+-----+-----+-----+-----+-----+
+/// | P1  | P2  | P3  | P4  | P5  | P6  | P7  | P8  |  Permutation Sets
+/// +-----+-----+-----+-----+-----+-----+-----+-----+
+///            . . . (Process repeats)
+///
+/// Legend:
+/// [Hx] : Hash value in SIMD lane x
+/// Px   : Permutation set x, where h'(x) = (a * x + b) % p
+/// Mx   : Running minimum hash value for permutation set x
+/// ```
+#[inline(always)]
+fn simd_permute_and_min_batch(
+    initial_hash: SimdU64,
+    perm_a: &[SimdU64],
+    perm_b: &[SimdU64],
+    min_hashes: &mut [SimdU64],
+) {
+    let mut rotated_hash = initial_hash;
+
+    debug_assert_eq!(
+        perm_a.len(),
+        perm_b.len(),
+        "Permutation vectors must have the same length"
+    );
+
+    debug_assert_eq!(
+        min_hashes.len(),
+        perm_a.len(),
+        "Minimum hash values must have the same length as the permutation vectors"
+    );
+
+    let perm_a = perm_a.iter();
+    let perm_b = perm_b.iter();
+    let min_hashes = min_hashes.iter_mut();
+
+    for ((coefficient_a, coefficient_b), current_min_hash) in perm_a.zip(perm_b).zip(min_hashes) {
+        // Apply permutations and update minimum hash values for each SIMD lane
+        for _ in 0..SIMD_LANES {
+            let permuted_hash =
+                compute_fast_simd_remainder(rotated_hash * coefficient_a + coefficient_b);
+            *current_min_hash = permuted_hash.simd_min(*current_min_hash);
+            // Rotate the hash vector left by 1 element. This ensures that each SIMD lane
+            // processes a different permutation of the initial hash in subsequent iterations,
+            // effectively computing multiple hash permutations in parallel.
+            rotated_hash = rotated_hash.rotate_elements_left::<1>();
```

Comment:
This was our initial implementation when we wrote this kernel and found that it was much slower on x86 machines on AWS where this is most commonly ran. We favored the `splat` approach due to that. Let's keep the algorithm the same?

Reply:
https://github.com/Eventual-Inc/Daft/blob/0c390074f55af3ed40238ae2013c8d1fac713af5/src/daft-minhash/src/minhash.rs#L30